### PR TITLE
Remove most specialization use in serialization

### PR DESCRIPTION
**Pull request by matthewjasper**
Switching from specialization to min_specialization in the compiler made the unsoundness of how we used these traits pretty clear. This changes how the `Encodable` and `Decodable` traits work to be more friendly for types need a `TyCtxt` to deserialize.

The alternative design of having both `Encodable` and `TyEncodable` traits was considered, but doesn't really work because the following impls would conflict:

```
impl<E: Ecodable> TyEncodable for Encodable
impl<E: TyEcodable> TyEncodable for [E]
```

## How-to guide

- `Rustc(De|En)codable` is now spelled `Ty(De|En)coable` in `rustc_middle`, `Metadata(En|De)codable` in `rustc_metadata` where needed, and `(De|En)codable` everywhere else.
- Manual implementations of `(De|En)codable` shouldn't be much different. 
- If you're adding a new interned type that needs to be en/decodable then the simplest thing way to handle this is:
    - Have the type be a wrapper around a reference to the interned data (i.e. do what `ty::Predicate` does, and not what all of the other interned types do)
    - Derive `Ty(En|De)codable` on the inner type
    - Implement `Encodable<impl TyEncoder>` by forwarding to the inner type.
    - Implement `Decodable<impl TyDecoder>` by decoding the inner type and then creating the wrapper around that (using the `tcx` from the decoder as needed).

cc @rust-lang/compiler for opinions on this change 
r? @oli-obk

### CODE REVIEW

File: src/librustc_serialize/serialize.rs [hunk 1]
```diff
@@ -380,282 +379,310 @@ pub trait Decoder {
...
-    fn decode<D: Decoder>(d: &mut D) -> Result<Self, D::Error>;
+/// Trait for types that can be serialized
+///
+/// This can be implemented using the `Encodable`, `TyEncodable` and
+/// `MetadataEncodable` macros.
+///
+/// * `Encodable` should be used in crates that don't depend on
+///   `librustc_middle`.
+/// * `TyEncodable` should be used for types that are only serialized in crate
+///   metadata or the incremental cache, except for simple enums.where
```
Comment (oli-obk) [inline src/librustc_serialize/serialize.rs:390]: why except for simple enums, and why does this rule not apply to `MetadataEncodable`? Also there's a trailing `where` here.

File: src/librustc_serialize/serialize.rs [hunk 2]
```diff
@@ -380,282 +379,310 @@ pub trait Decoder {
...
-    fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+/// Trait for types that can be deserialized
+///
+/// This can be implemented using the `Decodable`, `TyDecodable` and
+/// `MetadataDecodable` macros.
+///
+/// * `Decodable` should be used in crates that don't depend on
+///   `librustc_middle`.
+/// * `TyDecodable` should be used for types that are only serialized in crate
+///   metadata or the incremental cache, except for simple enums.where
```
Comment (oli-obk) [inline src/librustc_serialize/serialize.rs:405]: same as above

File: src/librustc_serialize/serialize.rs [hunk 3]
```diff
@@ -380,282 +379,310 @@ pub trait Decoder {
...
+
+impl<S: Encoder> Encodable<S> for usize {
+    fn encode(&self, s: &mut S) -> Result<(), S::Error> {
         s.emit_usize(*self)
     }
 }
 
-impl Decodable for usize {
-    fn decode<D: Decoder>(d: &mut D) -> Result<usize, D::Error> {
+impl<D: Decoder> Decodable<D> for usize {
```
Comment (oli-obk) [inline src/librustc_serialize/serialize.rs:418]: preexisting, but it seems to me like we should have a macro for all these trivial impls in this file.

File: src/librustc_serialize/serialize.rs [hunk 4]
```diff
@@ -380,282 +379,310 @@ pub trait Decoder {
...
-    fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+impl<S: Encoder> Encodable<S> for str {
+    fn encode(&self, s: &mut S) -> Result<(), S::Error> {
         s.emit_str(self)
     }
 }
 
-impl Encodable for String {
-    fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+impl<S: Encoder> Encodable<S> for &str {
```
Comment (oli-obk) [inline src/librustc_serialize/serialize.rs:574]: I'm guessing a general impl for `&T where T: Encodable<S>` will conflict with something?

**Review by oli-obk — APPROVED** [review]
>     * If you're adding a new interned type that needs to be en/decodable then the simplest thing way to handle this is:

maybe include this info in the module/crate docs of rustc_serialize

I didn't check the derive code in detail, but since it's working I'm fine merging it.

r=me with the mentioned documentation changes and CI passing

**Comment by nikomatsakis** [comment]
This seems like an MCP-worthy change.

**Comment by Aaron1011** [comment]
Let's see what the perf impact is:

@bors try @rust-timer queue

File: src/librustc_serialize/serialize.rs (hunk 4 above)
Comment (matthewjasper) [reply src/librustc_serialize/serialize.rs:574]: It conflicts with the impl for `Ty`.

**Comment by mark-i-m** [comment]
Should this be added to rustc-dev-guide in some form? We don't really have any documentation about serialization...

**Comment by nikomatsakis** [comment]
Perf results appear neutral to positive.

**Comment by nikomatsakis** [comment]
I do think we should have this setup documented in the rustc-dev-guide, yes.

**Comment by matthewjasper** [comment]
I've opened an MCP and a PR to document serialization in rustc in the guide.

**Comment by eddyb** [comment]
cc #36588

File: src/librustc_macros/src/serialize.rs [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        [vi] => {
+            let construct = vi.construct(|field, index| decode_field(field, index, true));
+            let n_fields = vi.ast().fields.len();
+            quote! {
+                ::rustc_serialize::Decoder::read_struct(
+                    __decoder,
+                    #ty_name,
+                    #n_fields,
+                    |__decoder| { ::std::result::Result::Ok(#construct) },
+                )
```
Comment (eddyb) [inline src/librustc_macros/src/serialize.rs:53]: It's neat that this works with `rustc_serialize`, I guess. For #36588 I expected we'd remove all of the features we don't need for the opaque encoding but I don't have anything against being comprehensive.

**Comment by matthewjasper** [comment]
@bors p=1 rollup=iffy
This will break code that adds new `Rustc(De|En)codable` uses.

**Comment by matthewjasper** [comment]
@oli-obk is this pr still r=you?

**Comment by oli-obk** [comment]
yes, unless you do a change beyond getting it to work again after a rebase, assume the `r=me` stands

**Comment by oli-obk** [comment]
@bors p=6

new rules around rollups say that bitrotty things get a priority above 5, otherwise rollups can still go ahead of them

**Comment by ecstatic-morse** [comment]
Final perf results show a [very slight regression](https://perf.rust-lang.org/compare.html?start=95879ad96104afa584e7aec7806cec1d0bd84116&end=668a34e0f438d4a950b9440239656d6755ad963c&stat=instructions:u) in instruction counts unlike the earlier [perf run](https://github.com/rust-lang/rust/pull/73851#issuecomment-651410644), which showed a slight improvement. Is this actionable @matthewjasper?

**Comment by matthewjasper** [comment]
I don't think there's anything obvious to do for this. Maybe profiling would show something.

**Pull request merged**